### PR TITLE
Support hard links in tarballs

### LINKED ISSUE #10395: fetchTarball does not support hard links
**Describe the bug**

A custom derivation using `fetchTarball` suddenly stopped building.

<details> <summary>Full error log</summary>

```
evaluating file '<nix/derivation-internal.nix>'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/minver.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/top-level/impure.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/top-level/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/fixed-points.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/trivial.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/lists.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/stdenv/booter.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/stdenv/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/systems/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/attrsets.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/systems/inspect.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/systems/architectures.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/systems/parse.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/strings.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/types.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/systems/platforms.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/modules.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/top-level/config.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/options.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/stdenv/linux/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/stdenv/linux/bootstrap-tools/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/top-level/stage.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/stdenv/adapters.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/build-support/trivial-builders/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/top-level/splice.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/top-level/by-name-overlay.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/top-level/all-packages.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/top-level/aliases.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/stdenv/generic/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/customisation.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/build-support/bintools-wrapper/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/stdenv/generic/make-derivation.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/asserts.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/tools/misc/binutils/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/servers/x11/xorg/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/servers/x11/xorg/overrides.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/build-support/cc-wrapper/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/stdenv/generic/check-meta.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/meta.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/stdenv/linux/bootstrap-files/x86_64-unknown-linux-gnu.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/stdenv/linux/bootstrap-files/i686-unknown-linux-gnu.nix'
evaluating file '<nix/fetchurl.nix>'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/tools/misc/patchelf/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/all.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/versions.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/versions.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/common/libgcc.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/common/builder.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/common/checksum.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/glibc/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/glibc/common.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/os-specific/linux/kernel-headers/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/gmp/6.x.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/isl/0.20.0.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/isl/generic.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/libmpc/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/mpfr/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/licenses.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/lib/systems/doubles.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/common/meta.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/maintainers/team-list.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/zlib/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/build-support/fetchurl/boot.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/gnu-config/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/gettext/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/shells/bash/5.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/shells/bash/bash-5.2-patches.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/tools/parsing/bison/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/tools/misc/gnum4/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/build-support/fetchurl/mirrors.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/interpreters/perl/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/interpreters/perl/intepreter.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/libxcrypt/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/tools/compression/xz/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/build-support/expand-response-params/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/common/extra-target-flags.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/common/dependencies.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/common/configure-flags.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/common/platform-flags.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/tools/misc/texinfo/7.0.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/tools/misc/texinfo/common.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/tools/system/which/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/patches/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/build-support/nuke-references/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/common/pre-configure.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/compilers/gcc/common/strip-attributes.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/interpreters/python/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/interpreters/python/cpython/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/interpreters/python/passthrufun.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/tools/compression/bzip2/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/expat/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/libffi/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/tools/misc/autoconf/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/tools/misc/automake/automake-1.16.x.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/tools/misc/libtool/libtool2.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/tools/misc/file/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/build-support/pkg-config-wrapper/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/tools/misc/pkg-config/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/tools/misc/autoconf-archive/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/interpreters/python/setup-hook.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/libidn2/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/libunistring/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/tools/misc/coreutils/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/development/libraries/acl/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/tools/text/gnugrep/default.nix'
evaluating file '/nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/tools/text/gawk/default.nix'
downloading 'https://www.gaisler.com/anonftp/bcc2/bin/bcc-2.2.1-gcc-linux64.tar.xz'...
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:9:12:
            8|
            9|   strict = derivationStrict drvAttrs;
             |            ^
           10|

       … while evaluating derivation 'gaisler-bcc-2.2.1'
         whose name attribute is located at /nix/store/sa8wlkyv4zvzkrx2a034ikfsn7iaj2k3-nixos-23.11-23.11/nixos-23.11/pkgs/stdenv/generic/make-derivation.nix:348:7

       … while evaluating attribute 'src' of derivation 'gaisler-bcc-2.2.1'
         at /tmp/nix-bug/toolchain.nix:11:7:
           10|       name = "gaisler-bcc-${version}";
           11|       src = builtins.fetchTarball {
             |       ^
           12|         url = tarballUrl;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: file 'bcc-2.2.1-gcc/bin/sparc-gaisler-elf-gcc' in tarball has unsupported file type

```
</details>

When importing the nix file using the repl, the error seems to appear after downloading a few MBs:

```
nix-repl> import ./toolchain.nix { pkgs = import <nixos-23.11> {}; }
{
[15.1/355.2 MiB DL] downloading 'https://www.gaisler.com/anonftp/bcc2/bin/bcc-2.2.1-gcc-linux64.tar.xz'«error: file 'bcc-2.2.1-gcc/bin/sparc-gaisler-elf-gcc' in tarball has unsupported file type»;
}
```

---

I was able to reproduce the error on both 23.11 and unstable channels, in two separate pcs:

- PC 1 (Arch 6.6.23 LTS):
  - `Linux hera 6.6.23-1-lts #1 SMP PREEMPT_DYNAMIC Wed, 27 Mar 2024 07:47:20 +0000 x86_64 GNU/Linux`
  - `nix-env (Nix) 2.21.1`

- PC 2 (Arch 6.8.2):
  - `Linux artemis 6.8.2-arch2-1 #1 SMP PREEMPT_DYNAMIC Thu, 28 Mar 2024 17:06:35 +0000 x86_64 GNU/Linux`
  - `nix-env (Nix) 2.21.1`

I couldn't reproduce it on the following computer:
- PC 3 (Ubuntu 22.04 LTS):
  - `Linux TITAN 6.5.0-26-generic #26~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Mar 12 10:22:43 UTC 2 x86_64 x86_64 x86_64 GNU/Linux`
  - `nix-env (Nix) 2.19.2`

---

You can find the MRE down below:

<details> <summary>toolchain.nix</summary>

```nix
{ pkgs }:

let
  version = "2.2.1";
  hash = "097z15zshr4711rzfr8ccwl8vzpxp7da573p89qmngxzrwxr4kwl";
  tarballUrl = "https://www.gaisler.com/anonftp/bcc2/bin/bcc-${version}-gcc-linux64.tar.xz";

  bccToolchain = pkgs.stdenv.mkDerivation
    {
      name = "gaisler-bcc-${version}";
      src = builtins.fetchTarball {
        url = tarballUrl;
        sha256 = hash;
      };

    };

in { inherit bccToolchain; }
```
</details>

Feel free to ping me if I missed any details.

### PR DESCRIPTION
**Pull request by edolstra**
# Motivation

Fixes #10395.

# Context

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libfetchers/git-utils.cc [hunk 1]
```diff
@@ -909,6 +909,50 @@ struct GitFileSystemObjectSinkImpl : GitFileSystemObjectSink
...
+        auto pathComponents = tokenizeString<std::vector<std::string>>(path, "/");
+        if (!prepareDirs(pathComponents, false)) return;
+
+        auto relTarget = CanonPath(path).parent()->makeRelative(target);
+
+        auto dir = pendingDirs.rbegin();
+
+        // For each ../ component at the start, go up one directory.
+        std::string_view relTargetLeft(relTarget);
+        while (hasPrefix(relTargetLeft, "../")) {
```
Comment (roberth) [inline src/libfetchers/git-utils.cc:923]: I was trying to see whether all `..`'s are guaranteed to come first when I realised: Is this logic needed? The test file doesn't have any relative paths, suggesting that we could just look up the files directly from the root.

While I like the idea of being lenient towards inputs, we run the risk that a bug here will go unnoticed for a long time, and putting reproducibility at stake.

As far as I can see (I haven't excluded a libarchive quirk), I think we could either
- Figure out a way to make `tar` produce relative hard links (idk if it can)
- Modify the tarball by hand to contain relative paths, see how GNU and BSD tar behave, and use the modified tarball in the test
- ~~Unit testing~~ would potentially test the wrong behavior
- Check that `..` does not occur in any of our path components and `throw UnimplementedError("Nix does not support tarballs with relative hard links yet. Please open an issue.")` if it does. We may never have to implement it.

File: tests/functional/tarball.sh [hunk 2]
```diff
@@ -59,3 +59,12 @@ test_tarball() {
 test_tarball '' cat
 test_tarball .xz xz
 test_tarball .gz gzip
+
+# Test hard links.
```
Comment (roberth) [inline tests/functional/tarball.sh:63]: Confirmed. Let's give some context:

```suggestion
# Test hard links.
# All entries in tree.tar.gz refer to the same file, and all have the same inode when unpacked by GNU tar
# We don't preserve the hard links, because that's an optimization we think is not worth the complexity,
# so we only make sure that the contents are copied correctly.
```

**Review by roberth — CHANGES_REQUESTED** [review]
I can't entirely rule out significant problems, as the relative path functionality appears to be untested; see comment.

File: src/libfetchers/git-utils.cc (hunk 1 above)
Comment (edolstra) [reply src/libfetchers/git-utils.cc:923]: The path returned by libarchive is always absolute (or to be precise, relative to the root of the archive). The relative path here comes from `CanonPath::makeRelative()`, which does indeed always put the `..` elements at the start. The conversion to a relative path is just to deal with the `git_tree_builders` that are in progress (we can't just look up the path from the start of the root, since some parent directories may not have finished yet).

File: src/libfetchers/git-utils.cc [hunk 3]
```diff
@@ -909,6 +909,50 @@ struct GitFileSystemObjectSinkImpl : GitFileSystemObjectSink
...
+    void createHardlink(const Path & path, const CanonPath & target) override
+    {
+        auto pathComponents = tokenizeString<std::vector<std::string>>(path, "/");
+        if (!prepareDirs(pathComponents, false)) return;
+
+        auto relTarget = CanonPath(path).parent()->makeRelative(target);
+
+        auto dir = pendingDirs.rbegin();
+
+        // For each ../ component at the start, go up one directory.
```
Comment (roberth) [inline src/libfetchers/git-utils.cc:921]: These comments would be helpful:

```suggestion
        // We can't just look up the path from the start of the root, since
        // some parent directories may not have finished yet, so we compute
        // a relative path that helps us find the right git_tree_builder or object.
        auto relTarget = CanonPath(path).parent()->makeRelative(target);

        auto dir = pendingDirs.rbegin();

        // For each ../ component at the start, go up one directory.
        // CanonPath::makeRelative() always puts all .. elements at the start,
        // so they're all handled by this loop:
```

**Comment by Ericson2314** [comment]
Do we need a new sink class? I suspect that every sink we have today can support this. The only other ones I know of from memory are writing to the OS filesystem and writing to the memory source accessor, and they both should support it. (The memory one can just copy.)

**Comment by roberth** [comment]
> Do we need a new sink class? I suspect that every sink we have today can support this.

I find the separation insightful.

**Comment by Ericson2314** [comment]
I guess it doesn't really matter.

My thinking was that since we don't support mutation of files, `ln` = `cp` = `cp --reflink` i.e. its not an extension to the data model so much as an expedient way to make copies.

**Comment by edolstra** [comment]
My thinking is that if you're just parsing a NAR, the interface shouldn't require you to handle features that don't exist in NARs.

File: tests/functional/tree.tar.gz
Comment (roberth) [inline tests/functional/tree.tar.gz:1]: This file is hard to understand, making it hard to see whether the corner cases have been covered.
That's not to say that we shouldn't have this; we should, but it'd be very helpful to also have a unit test, which will be easier to understand, review and improve.

Comment (edolstra) [reply tests/functional/tree.tar.gz:1]: Hm, what is hard to understand about it?

We could get rid of `tree.tar.gz` by generating it in the test script, though we have to make sure the `tar` version used generates hard links. Not sure how a unit test would help in this case.

Comment (roberth) [reply tests/functional/tree.tar.gz:1]: It's hard to read a raw tarball. I've looked at the strings inside, to get an idea of what's going on, but without looking at the hexdump or using some sort of parsing tool, I can only guess what is approximately going on.

A unit test isolates us from that complexity, or opacity. It provides a clearly readable "script" of what we can expect to happen on the Nix side when a tarball is read. 

> We could get rid of `tree.tar.gz` by generating it

We should not remove the file, because it's our only "end to end" test and we know that it certainly contains hardlinks, even if it's not super easy to figure that out.
Generating it would solve part of the problem, but suffers from the problem you describe. It does make it easier to understand and modify.
A unit test wouldn't be "end to end", but that's mostly covered by keeping the tarball file.
The way it helps is to also make it easier to understand and modify, and besides that, it may let us construct corner cases more easily without other, complicated software getting in our way. Another benefit of a unit test is that it provides an abstracted view of the problem, which is very useful for the next person, whether that's for debugging, reviewing, ... (also it's the reason to keep the tarball file around)

**Comment by roberth** [comment]
I've opened https://github.com/DeterminateSystems/nix-src/pull/1 to add the first unit tests to this, but I haven't got around to completing the testing.
